### PR TITLE
rewrite atomFamily using Map

### PR DESCRIPTION
**Pull request by likern**
I expect this patch should improve performance when using `atomFamily`

### CODE REVIEW

**Comment by dai-shi** [comment]
This is nice! I don't remember why I didn't use `Map` (maybe confused with `WeakMap` which only works with objects.)

(BTW, are you in pmndrs discord server?)

**Comment by likern** [comment]
> This is nice! I don't remember why I didn't use `Map` (maybe confused with `WeakMap` which only works with objects.)
> 
> (BTW, are you in pmndrs discord server?)

Yes, I'm there in **jotai** group as a likern user.

File: src/utils/atomFamily.ts [hunk 1]
```diff
@@ -61,45 +60,52 @@ export function atomFamily<Param, Value, Update extends never = never>(
...
       initializeRead(param),
       initializeWrite && initializeWrite(param)
     ) as AtomType
-    atoms.unshift([param, newAtom, Date.now()])
+    atoms.set(param, [newAtom, Date.now()])
     return newAtom
   }
+
   createAtom.remove = (p: Param) => {
-    const index = atoms.findIndex((x) => x[0] === p)
```
Comment (dai-shi) [inline src/utils/atomFamily.ts:88]: Just noticed my original implementation doesn't take care of `areEqual`.
So, this is not a regression in this PR, but we should use areEqual here too to be consistent.

**Review by dai-shi — APPROVED** [review]
Nice work! @likern 

Also a big thank to @merisbahti who created the test #101 !

File: src/utils/atomFamily.ts (hunk 1 above)
Comment (dai-shi) [reply src/utils/atomFamily.ts:88]: #177 follow up issue.

**Pull request merged**